### PR TITLE
feat(challenger): Challenger game scanner

### PR DESCRIPTION
**Pull request by jackchuma**
### What changed? Why?
Adds the GameScanner component to the challenger service, which monitors the DisputeGameFactory contract for dispute games that need validation. The scanner filters games by type, in-progress status, and unchallenged state (zero zkProver), returning candidate games for downstream processing. Also extends the AggregateVerifierClient trait with three new methods (status, zk_prover, starting_block_number) needed for per-game queries.

### How has it been tested?
All 30 tests pass (7 new scanner tests covering happy path, all filter conditions, edge cases, and error resilience), clippy clean.

### Notes to reviewers
Purely additive — no existing behavior is modified. The scanner is stateless across restarts (lookback window recomputed on startup). Uses trait objects (Arc<dyn AggregateVerifierClient>) for testability. Individual game query failures are logged and skipped to stay resilient against transient RPC errors.

### Change management
type=routine
risk=low
impact=sev5

### Backwards Compatibility
backwards_compatible=true

automerge=false

### CODE REVIEW

**Comment by cb-heimdall** [comment]
<h1>✅ Heimdall Review Status</h1>
<table>
<tr>
<th>Requirement</th>
<th>Status</th>
<th>More Info</th>
</tr>
<tr>
<td>Reviews</td>
<td>
✅
<code>1/1</code>
</td>
<td>
<details>
<summary>Denominator calculation</summary>
<table>
<tr>
<td>
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><code>1</code> if user is bot</td>
<td>0</td>
</tr>
<tr>
<td><code>1</code> if user is external</td>
<td>0</td>
</tr>
<tr>
<td><code>2</code> if repo is sensitive</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>.codeflow.yml</code>
</td>
<td>1</td>
</tr>
<tr>
<td>
Additional review requirements
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><strong>Max</strong></td>
<td><strong>0</strong></td>
</tr>
</table>
</details>
</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>CODEOWNERS</code>
</td>
<td>0</td>
</tr>
<tr>
<td>Global minimum</td>
<td>0</td>
</tr>
<tr>
<td><strong>Max</strong></td>
<td>
<strong>
1
</strong>
</td>
</tr>
</table>
</details>
</td>
<td>1</td>
</tr>
<tr>
<td><code>1</code> if commit is unverified</td>
<td>0</td>
</tr>
<tr>
<td><strong>Sum</strong></td>
<td><strong>1</strong></td>
</tr>
</table>
</details>
</td>
</tr>
</table>

File: crates/proof/challenge/src/scanner.rs [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+use crate::ChallengerMetrics;
+
+/// Game status indicating the dispute is still in progress.
+pub const STATUS_IN_PROGRESS: u8 = 0;
+
+/// Configuration for the game scanner.
+#[derive(Debug, Clone)]
+pub struct ScannerConfig {
+    /// Game type ID to filter for.
+    pub game_type: u32,
```
Comment (refcell) [inline crates/proof/challenge/src/scanner.rs]: only supporting 1 type?

File: crates/proof/challenge/src/scanner.rs [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+    pub proxy: Address,
+    /// The output root claimed by this game.
+    pub root_claim: B256,
+    /// The L2 block number of this game.
+    pub l2_block_number: u64,
+    /// The parent game's factory index.
+    pub parent_index: u32,
+    /// The starting block number for this game.
+    pub starting_block_number: u64,
+}
```
Comment (refcell) [inline crates/proof/challenge/src/scanner.rs:42]: is this type not already defined by the contract bindings? Feels like there should already be a DisputeGame type but maybe I'm wrong.

File: crates/proof/challenge/src/scanner.rs [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+        for i in start..=end {
+            match self.evaluate_game(i).await {
+                Ok(Some(candidate)) => candidates.push(candidate),
+                Ok(None) => {}
+                Err(e) => {
+                    warn!(error = %e, index = i, "failed to query game, skipping");
+                    lowest_error = lowest_error.map_or(Some(i), |prev| Some(prev.min(i)));
+                }
+            }
+        }
```
Comment (refcell) [inline crates/proof/challenge/src/scanner.rs]: iterate in parallel?

Comment (refcell) [reply crates/proof/challenge/src/scanner.rs]: https://docs.rs/rayon/latest/rayon/iter/trait.ParallelIterator.html

File: crates/proof/challenge/src/scanner.rs [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+pub struct ScannerConfig {
+    /// Game type ID to filter for.
+    pub game_type: u32,
+    /// Number of past games to scan on startup (lookback window).
+    pub lookback_games: u64,
+}
+
+/// A dispute game that has been identified as a candidate for challenge.
+#[derive(Debug, Clone)]
+pub struct CandidateGame {
```
Comment (refcell) [inline crates/proof/challenge/src/scanner.rs:33]: One thing to note here is there could be a case where a game is created and then the l1 block is re-orged so the candidate game doesn't exist anymore between scanning and challenging the game. We just don't want to cause the service to panic if that is the case.

**Review by refcell — COMMENTED** [review]
Few comments

File: crates/proof/challenge/src/test_utils.rs [hunk 5]
```diff
@@ -0,0 +1,475 @@
...
+use async_trait::async_trait;
+use base_proof_contracts::{
+    AggregateVerifierClient, ContractError, DisputeGameFactoryClient, GameAtIndex, GameInfo,
+};
+
+use crate::scanner::{GameScanner, ScannerConfig};
+
+/// Per-game state for the mock verifier.
+#[derive(Debug, Clone)]
+pub(crate) struct MockGameState {
```
Comment (0x00101010) [inline crates/proof/challenge/src/test_utils.rs]: CLAUDE.md violation: `pub(crate)` types in test_utils.rs instead of `pub`

CLAUDE.md mandates: "All structs, types, enums, and functions within modules should be pub … No private or pub(crate) types." Four structs in test_utils.rs are declared pub(crate) instead of pub: MockGameState (line 15), MockDisputeGameFactory (line 27), MockAggregateVerifier (line 55), and ErrorOnIndexFactory (line 129). Since the module itself is gated by #[cfg(test)] at lib.rs:23, making the types pub does not leak them into the production API — the module visibility already gates access. The types should be pub to conform to the rule.

Same for all other below

File: crates/proof/challenge/src/scanner.rs [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+use base_proof_contracts::{
+    AggregateVerifierClient, DisputeGameFactoryClient, GameAtIndex, GameInfo,
+};
+use eyre::Result;
+use tracing::{debug, info, warn};
+
+use crate::ChallengerMetrics;
+
+/// Game status indicating the dispute is still in progress.
+pub const STATUS_IN_PROGRESS: u8 = 0;
```
Comment (0x00101010) [inline crates/proof/challenge/src/scanner.rs]: CLAUDE.md violation: `STATUS_IN_PROGRESS` exported as a loose constant instead of an associated constant on a type

CLAUDE.md states: "Prefer placing functions as methods on a type (even a unit struct) rather than as bare functions, so the public API exports types, not loose functions." STATUS_IN_PROGRESS is defined as a module-level pub const at scanner.rs:25 and re-exported as a loose item from lib.rs:18. The existing pattern in this crate places constants as associated consts on types (e.g., ChallengerMetrics::INFO, ChallengerMetrics::UP in metrics.rs:7-10). This constant should be an associated constant on GameScanner (or similar type) to follow the established convention and the CLAUDE.md rule.

File: crates/proof/challenge/src/test_utils.rs [hunk 7]
```diff
@@ -0,0 +1,475 @@
...
+#[async_trait]
+impl AggregateVerifierClient for MockAggregateVerifier {
+    async fn game_info(&self, game_address: Address) -> Result<GameInfo, ContractError> {
+        self.games
+            .get(&game_address)
+            .map(|s| s.game_info.clone())
+            .ok_or_else(|| ContractError::Validation(format!("unknown game {game_address}")))
+    }
+
+    async fn status(&self, game_address: Address) -> Result<u8, ContractError> {
```
Comment (0x00101010) [inline crates/proof/challenge/src/test_utils.rs:71]: not related to this PR, but @danyalprout @refcell, what's our stance on eyre::Result instead of std Result? personally not a big fan of typed errors, makes things too verbose and we don't really check the type struct probably

**Review by refcell — APPROVED** [review]
lgtm
